editoast/editoast_models/src/db_connection_pool.rs

```diff
@@ -17,21 +21,81 @@ use openssl::ssl::SslMethod;
 use openssl::ssl::SslVerifyMode;
 use url::Url;
 
-#[cfg(feature = "testing")]
 use tokio::sync::OwnedRwLockWriteGuard;
-#[cfg(feature = "testing")]
 use tokio::sync::RwLock;
 
-use super::DbConnection;
 use super::DbConnectionPool;
+use super::DieselConnection;
 
 pub type DbConnectionConfig = AsyncDieselConnectionManager<AsyncPgConnection>;
 
-#[cfg(feature = "testing")]
-pub type DbConnectionV2 = OwnedRwLockWriteGuard<Object<AsyncPgConnection>>;
+#[derive(Clone)]
+pub struct DbConnection {
+    inner: Arc<RwLock<Object<AsyncPgConnection>>>,
+}
 
-#[cfg(not(feature = "testing"))]
-pub type DbConnectionV2 = Object<AsyncPgConnection>;
+pub struct WriteHandle {
+    guard: OwnedRwLockWriteGuard<Object<AsyncPgConnection>>,
+}
+
+impl DbConnection {
+    pub fn new(inner: Arc<RwLock<Object<AsyncPgConnection>>>) -> Self {
+        Self { inner }
+    }
+
+    pub async fn write(&self) -> WriteHandle {
+        WriteHandle {
+            guard: self.inner.clone().write_owned().await,
+        }
+    }
+
+    pub async fn transaction<'a, R, E, F>(self, callback: F) -> std::result::Result<R, E>
+    where
+        F: FnOnce(Self) -> ScopedBoxFuture<'a, 'a, std::result::Result<R, E>> + Send + 'a,
+        E: From<diesel::result::Error> + Send + 'a,
+        R: Send + 'a,
+    {
+        use diesel_async::TransactionManager as _;
+
+        type TxManager = <AsyncPgConnection as AsyncConnection>::TransactionManager;
+
+        {
+            let mut handle = self.write().await;
+            TxManager::begin_transaction(handle.deref_mut()).await?;
+        }
+
+        match callback(self.clone()).await {
+            Ok(result) => {
+                let mut handle = self.write().await;
+                TxManager::commit_transaction(handle.deref_mut()).await?;
+                Ok(result)
+            }
+            Err(callback_error) => {
+                let mut handle = self.write().await;
+                match TxManager::rollback_transaction(handle.deref_mut()).await {
+                    Ok(()) | Err(diesel::result::Error::BrokenTransactionManager) => {
+                        Err(callback_error)
+                    }
+                    Err(rollback_error) => Err(rollback_error.into()),
+                }
+            }
+        }
+    }
```

Comment:
I believe this piece of code takes inspiration from the existing implementation inside `diesel` crate (see [`TransactionManager::transaction()`](https://docs.rs/diesel/2.1.6/src/diesel/connection/transaction_manager.rs.html#50-71)). It's probably a good idea to keep a link to it and refer to it so the next developer that needs to maintain this function know it should also take inspiration on the `diesel` crate. Maybe also add a comment on the reason why we have to reimplement it ourselves and we can't use the existing one.

Reply:
Proposition of documentation in https://github.com/OpenRailAssociation/osrd/pull/8712/commits/9fef9a6a13826c3c8cf2c5ba809dc1aa61b2314c. @leovalais What do you think?